[FrameworkBundle] Functional testing system refactoring

fduch opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | tiny (for tests)
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #14590
| License       | MIT
| Doc PR        | no

Please refer to initial discussions here: #14590. 

There is a overview of the changes in a nutshell:

- Fix fetching of the test kernel class from the KERNEL_DIR that [returned only basename of the first file ends with *Kernel](https://github.com/symfony/symfony/blob/3.2/src/Symfony/Bundle/FrameworkBundle/Test/KernelTestCase.php#L128-L133) and not actually the class name and therefore could not be instantiated with `new` operator (probably because of this it was neccesary to [completely override getKernelClass](https://github.com/symfony/symfony/blob/3.2/src/Symfony/Bundle/SecurityBundle/Tests/Functional/WebTestCase.php#L36-L41) to explicitly specify kernel name FQCN)
- Add ability to specify explicitly test kernel FQCN on the phpunit configuration level to avoid kinds of collision (please see **Problem** section of #14590)
- Add ability to override test kernel environment and debug-mode on the phpunit configuration level for example to support integration testing with real application kernel but in different environments.
- Introduce TestKernelInterface and extract [FrameworkBundle's test kernel class](https://github.com/symfony/symfony/blob/3.2/src/Symfony/Bundle/FrameworkBundle/Tests/Functional/app/AppKernel.php) in TestKernelInterface's implementation with adoptation to be usefull in arbitrary bundles.
- Remove code duplication and scattered temp directories clean in FrameworkBundle, SecurityBundle, TwigBundle (moved to TestKernel usage)

fduch committed:
refactor functional testing system

fduch commented:
Tests are broken since during test suite run symfony/framework-bundle dependency for packages is installed in stable version and not in version provided by this PR

fduch commented:
Any thoughts on this? /cc @weaverryan @jakzal

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Test/KernelTestCase.php:
```diff
@@ -11,7 +11,10 @@
 namespace Symfony\Bundle\FrameworkBundle\Test;
 
+use Symfony\Bundle\FrameworkBundle\Tests\Functional\Kernel\TestKernel;
+use Symfony\Bundle\FrameworkBundle\Tests\Functional\Kernel\TestKernelInterface;
```
just wondering: are these two classes required for KernelTestCase to work correctly?
because then that'd need a fix: no classes from the "Tests" namespace can be required for making things in "Test" work.

fduch replied:
no, they definitely are not required for correct work of KernelTestCase. Moved them to Test namespace

nicolas-grekas commented:
rebase needed

fduch commented:
@nicolas-grekas  rebase done

stof commented on src/Symfony/Bundle/FrameworkBundle/Test/KernelTestCase.php:
```diff
@@ -115,23 +127,59 @@ protected static function getKernelClass()
-        $class = $file->getBasename('.php');
+            $classes = get_declared_classes();
+            require_once $file;
+            $newClasses = array_diff(get_declared_classes(), $classes);
```
this logic is broken if the file was already loaded previously, as it won't declare a new class

fduch replied:
Replaced with ClassFinder invocation

stof commented on src/Symfony/Bundle/FrameworkBundle/Test/KernelTestCase.php:
```diff
@@ -165,10 +213,35 @@ protected static function createKernel(array $options = array())
 
-        return new static::$class(
+        if (isset($_SERVER['KERNEL_ENV'])) {
+            $options['environment'] = $_SERVER['KERNEL_ENV'];
```
options passed explicitly as argument must win over env variables IMO, otherwise this becomes a WTF for people passing options explicitly

fduch replied:
Replaced with current logic from 3.4 branch

stof commented on src/Symfony/Bundle/FrameworkBundle/Test/KernelTestCase.php:
```diff
@@ -165,10 +213,35 @@ protected static function createKernel(array $options = array())
+        }
+
+        if (isset($_SERVER['KERNEL_DEBUG'])) {
+            $options['debug'] = $_SERVER['KERNEL_DEBUG'];
```
this is broken, because env variables are always strings, while ``debug`` should be a boolean

fduch replied:
fixed

fduch commented on src/Symfony/Bundle/FrameworkBundle/Test/KernelTestCase.php:
```diff
@@ -115,23 +127,59 @@ protected static function getKernelClass()
-        $class = $file->getBasename('.php');
+            $classes = get_declared_classes();
+            require_once $file;
+            $newClasses = array_diff(get_declared_classes(), $classes);
```
Indeed there are no such cases for now but this can happen when several tests running in the same process share the same kernel directory. 

Should we tokenize found file here instead?

fduch commented on src/Symfony/Bundle/FrameworkBundle/Test/KernelTestCase.php:
```diff
@@ -165,10 +213,35 @@ protected static function createKernel(array $options = array())
 
-        return new static::$class(
+        if (isset($_SERVER['KERNEL_ENV'])) {
+            $options['environment'] = $_SERVER['KERNEL_ENV'];
```
This behavior can be useful for some kind of "integration" tests: for example you can take the external bundle and run it's test suite with your application kernel.

stof replied:
Passing options explicitly must still be respected. Otherwise, it is just a WTF.

fduch committed:
extract class finder in separate class and use it to find kernel classes in func tests. Make explicit kernel params more importants then server options

fabpot commented:
This one should be rebased as some features have been changed recently (which would perhaps simplify this PR as well).

fduch commented:
@fabpot, yes, i already know. 
I'll try to rebase and adopt it asap

nicolas-grekas commented:
@fduch d you think you can make it for the feat freeze? (end of week)

fduch commented:
@nicolas-grekas i'll try. Probably i'll not have enough time, so in this case i'll struggle with it on holidays in order to finish it for monday, ok?

fduch committed:
rebase & fixes

fduch commented:
Ok, i've changed the base branch to 3.4 and rebased.

Few comments:
1. Arranged specific testcase for [SubRequestsTest](https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/FrameworkBundle/Tests/Functional/SubRequestsTest.php#L18) instead of usage of Session one
2. Fixed kernel class resolving  from directory by usage of `ClassFinder` exposed from symfony/class-loader. This functionality should be removed in 4.0 since we do not support KERNEL_DIR in 4.0 anymore.
3. Adopted `getVarDir` usage.

stof commented on src/Symfony/Component/ClassLoader/ClassFinder.php:
```diff
@@ -0,0 +1,97 @@
+ *
+ * @author Gyula Sallai <salla016@gmail.com>
+ */
+class ClassFinder
```
Adding a new class here looks wrong to me, as the whole component is deprecated

fduch replied:
Yes, component is deprecated and will be removed in 4.0 just like the functionality of Kernel lookup inside KERNEL_DIR, where this code is used. 
I supposed that it probably worth to reuse the code (with introducing new class but without introducing any kind of functionality) instead of copy-pasting since this 2 code blocks have same roadmap.

fduch committed:
cs fixes

nicolas-grekas commented:
@fduch there are still many failures, the code needs to be updated to not use the features you deprecate anymore.

fduch commented:
@nicolas-grekas thanks for your reply)

If i see your point correctly, you are talking about [usage of deprecated getPhpUnitXmlDir during kernel class lookup near phpunit.xml.dist file](https://github.com/fduch/symfony/blob/fduch/func-tests-refactoring/src/Symfony/Bundle/FrameworkBundle/Test/KernelTestCase.php#L150-L154). 

The point here is that this code left conceptually untouched, but currently the whole method getKernelClass is subject to replace in implementations (probably because of problems i mentioned earlier), so original code is unreachable. Please refer to [FrameworkBundle WebTestCase](https://github.com/symfony/symfony/blob/3.4/src/Symfony/Bundle/FrameworkBundle/Tests/Functional/WebTestCase.php#L45-L49) or [SecurityBundle WebTestCase](https://github.com/symfony/symfony/blob/3.4/src/Symfony/Bundle/SecurityBundle/Tests/Functional/WebTestCase.php#L45-L50).

In proposed patch base of implementation of kernel class lookup is widely used again. Since KERNEL_CLASS is usually not specified (there is no need of this since base\Symfony\Bundle\FrameworkBundle\Test\TestKernel suites all the requisitions) the code with deprecations is evaluated. Omitting it means introduce BC break.

If we want to avoid deprecation notices, one possible workaround which i can see right now is to redundantly explicitly set  KERNEL_CLASS to \Symfony\Bundle\FrameworkBundle\Test\TestKernel in phpunit.xml.dist of FrameworkBundle, SecurityBundle, TwigBundle. 

WDYT?

nicolas-grekas commented:
I'm moving this PR to 4.1, we're already late in feat freeze.

fduch commented:
@nicolas-grekas it sounds like not so cool news (. 
Look, it is more then half a year in sum this PR was waiting for any kind of reactions (it also is waiting right now as well) which led to the necessity to perform two rebases with adoptation to the external changes each time. The last one was done in order to make it for the 3.4 feature freeze but review was postponed again.
Moreover there is part of this PR which is dedicated to 3+ functionality (it is mostly deprecated after PR was submitted but nevertheless it is part of 3+ and contains several problems i described above)  - and this part would be lost during new rebase against 4.x. 
If this one is conceptually approved should we probably do our best and squeeze into 3.4?

fabpot commented:
Not sure how to deal with this PR. There are so many PRs that would have needed more attention sooner, but unfortunately, the activity so intense on Symfony that we cannot be everywhere. Anyway, I'm very sorry that we are now in that situation, but I think that it's better to close it now.
